### PR TITLE
Update CustomFunction to allow optional parameters

### PR DESCRIPTION
**Pull request by alfaalex81**

### CODE REVIEW

File: src/NXP/Classes/CustomFunction.php [hunk 1]
```diff
@@ -32,28 +33,29 @@ class CustomFunction
...
             throw new IncorrectNumberOfFunctionParametersException($this->name);
         }
         $args = [];
-        for ($i = 0; $i < $this->places; $i++) {
+
+        for ($i = 0; $i < $this->placesOptional; $i++) {
             array_unshift($args, array_pop($stack)->value);
         }
-
+        $args = array_values(array_filter($args));
```
Comment (neonxp) [inline src/NXP/Classes/CustomFunction.php:58]: This line breaks tests. I suggest pass callback that strictly checks argument to null

File: src/NXP/Classes/CustomFunction.php [hunk 2]
```diff
@@ -32,28 +33,29 @@ class CustomFunction
...
-    public function __construct(string $name, callable $function, ?int $places = null)
-    {
+    public function __construct(string $name, callable $function, ?int $places = null) {
         $this->name = $name;
         $this->function = $function;
         if ($places === null) {
             $reflection = new ReflectionFunction($function);
-            $this->places = $reflection->getNumberOfParameters();
+            $this->places = $reflection->getNumberOfRequiredParameters();
+            $this->placesOptional = $reflection->getNumberOfParameters();
```
Comment (neonxp) [inline src/NXP/Classes/CustomFunction.php:42]: This returns optional and required args, not only optional.

File: src/NXP/Classes/CustomFunction.php [hunk 3]
```diff
@@ -32,28 +33,29 @@ class CustomFunction
...
-    {
+    public function execute(array &$stack): Token {
         if (count($stack) < $this->places) {
             throw new IncorrectNumberOfFunctionParametersException($this->name);
         }
         $args = [];
-        for ($i = 0; $i < $this->places; $i++) {
+
+        for ($i = 0; $i < $this->placesOptional; $i++) {
             array_unshift($args, array_pop($stack)->value);
```
Comment (neonxp) [inline src/NXP/Classes/CustomFunction.php:56]: No checks to empty stack (e.g. function has 3 required and 2 optional args, so in stack we may have 3 - 5 arguments)

**Review by neonxp — COMMENTED** [review]
Hello! Thank you for your contributions! But this PR has some problematic places.
I tried to make small fixes to pass exist tests, but now we need new tests with functions with optional arguments.
```
    public function __construct(string $name, callable $function, ?int $places = null, ?int $placesOptional = null) {
        $this->name = $name;
        $this->function = $function;
        if ($places === null) {
            $reflection = new ReflectionFunction($function);
            $this->places = $reflection->getNumberOfRequiredParameters();
            $this->placesOptional = $reflection->getNumberOfParameters() - $this->places; // num of last optional params
        } else {
            $this->places = $places;
            $this->placesOptional = $placesOptional;
        }
    }

    /**
     * @param Token[] $stack
     * @return Token
     */
    public function execute(array &$stack): Token {
        if (count($stack) < $this->places) {
            throw new IncorrectNumberOfFunctionParametersException($this->name);
        }
        $args = [];
        for ($i = 0; $i < $this->places; $i++) {
            array_unshift($args, array_pop($stack)->value);
        }
        for ($i = 0; $i < $this->placesOptional; $i++) {
            if (count($stack) != 0) {
                $val = $stack[count($stack)-1]->value;
                if ($val->type != Token::RightParenthesis) {
                    array_unshift($args, array_pop($stack)->value);
                }
            }
        }
        $args = array_values(array_filter($args, function($check) {
            return $check !== null;
        }));
        $result = call_user_func_array($this->function, $args);

        return new Token(Token::Literal, $result);
    }
}
```

**Comment by phpfui** [comment]
I am away for the weekend.  I may be able to take a look before I leave,
but I will for sure take a look Monday.  But certainly all tests must pass.

Bruce

**Comment by phpfui** [comment]
@alfaalex81,

Unfortunately I am going to have to reject this pull request. While I think it would be a good feature to have optional parameters on a function call, there is a bit more work that needs to be done to make this works correctly. And since I don't require this for my work, I am not interested in debugging it, but I will leave you with these comments in hope that you will work out all the issues and resubmit a PR.

First, all unit tests must run with no errors.  Looks like you are new to github, so welcome!  But you need to do a "composer install" followed by a "vendor/bin/phpunit" (reverse slashes for Windows) and that must run with no errors.

Second, additional unit tests (in tests directory) need to be added to insure the change works as expected.  I would create a function that takes multiple default arguments. I would multiply the first parameter by 10, then subtract the second parameter, and then repeat this pattern many times.  This will be a good test as it will insure you get the ordering of parameters correct.  And make sure to use parameters that insure the correct answer is returned without question.  Also add a test that will obviously fail.

And words to the wise: parsing can get complex, but if you have to add a lot of code into the parser to get this to work, you are probably doing it wrong and there should be a simpler solution.  I tried to do this once before in the old parser, but it got messy quickly.

As always, should you or any of your IM Force be caught or killed, the Secretary will disavow any knowledge of your actions.

**Pull request closed without merging**